src/easy/handler.rs

```diff
@@ -3128,6 +3128,20 @@ impl<H> Easy2<H> {
             Ok(list::from_raw(list))
         }
     }
+    /// Get the last http version number
+    ///
+    /// Corresponds to `CURLINFO_HTTP_VERSION` and may return an error if the
+    /// option isn't supported.
+    pub fn response_http_version(&self) -> Result<HttpVersion, Error> {
+        self.getopt_long(curl_sys::CURLINFO_HTTP_VERSION)
+            .map(|c| match c as i32 {
+                curl_sys::CURL_HTTP_VERSION_1_0 => HttpVersion::V10,
+                curl_sys::CURL_HTTP_VERSION_1_1 => HttpVersion::V11,
+                curl_sys::CURL_HTTP_VERSION_2_0 => HttpVersion::V2,
+                curl_sys::CURL_HTTP_VERSION_3 => HttpVersion::V3,
+                _ => unreachable!("TODO"),
```

Comment:
According to https://curl.se/libcurl/c/CURLINFO_HTTP_VERSION.html this only returns a select few codes, but should we map them all "just in case"? It can also return `0` in case the version cannot be determined, should we map that "none" to `Any` or wrap this in an `Option`? And should we wrap other faulty codes into a `Result` of sorts or panic on invalid library return values?

Sorry, still a bit new to this crate and `curl`, while adding this change on a tangent of a tangent :)

Reply:
We should assume that curl _could_ return anything here, and the last thing we want to do is panic when encountering an unknown result. Curl is often dynamically linked, meaning we don't have control over which version of libcurl is actually used. If old Rust code using old `curl-rust` code starts to run against a new libcurl version due to a system upgrade that starts returning a hypothetical "HTTP 4", then we:

1. For sure should not panic on any return value
2. If we can, return the underlying `i32` even though we don't understand it, allowing the caller to attempt to understand it themselves if they want to

One possible way of doing this would be to wrap the return value in a new enum that maybe looks something like this:

```rust
pub enum HttpVersionInfo {
    HttpVersion(HttpVersion),
    Unknown(i32),
}
```

Reply:
That conflicts a bit with the use of a `Result` type. I'd rather wrap the existing `Result<HttpVersion, CurlError>` with a `enum HttpVersionSpecificError { CurlError(CurlError), UnknownValue(i32) }` of sorts :)

Reply:
I ended up using your suggestion as `Unknown` isn't immediately invalid.

In another crate we've had discussions about a user hypothetically matching `Unknown(0x1337)`, and the crate being bumped with a semver-compatible release (or breaking release, doesn't really matter here) that maps the new `SuperDuperHttpVersion = 0x1337` value in this `#[non_exhaustive]` `enum`. The user will suddenly run into their mandatory `_ => {}` catch-all arm again. _Not_ giving them access to the value entices contributions and requests upstream to match the new value, and have less ambiguity inbetween.